### DISCUSSION TITLE
What are the type hints in Python models ?

### DISCUSSION THREAD

**Discussion by adrien-turiot-maxa** [dbt-labs/dbt-core#10054] [Q&A]
In a [python models](https://docs.getdbt.com/docs/build/python-models), we would like to have type hints. 
But the `dbt` argument does not have any definition, for example here:

```py
from snowflake.snowpark.session import Session

def model(dbt: ???, session: Session):
    ...
```

It looks like this is a generated class like `main_module.dbtObj` 🤔 

Is their a type available for this value ?

**Comment by ndrewwm** [comment]
I'm also curious about what would be the best fit here. I've found [this post](https://discourse.getdbt.com/t/dbt-python-model-dbt-py-best-practices/5204/2), but this is a bit unsatisfying.

**Reply by mateuscarestiato** [reply]
Hi @ndrewwm! The dbt object in Python models is a dynamically generated class at runtime, which is why there's no obvious public type to import directly. Here are the practical solutions:

Option 1: Use TYPE_CHECKING (recommended)

python
from future import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
from dbt.adapters.protocol import MacroProtocol
def model(dbt: "MacroProtocol", session):
df = dbt.ref("my_model")
...
This gives you IDE autocompletion without a runtime import error.

Option 2: Use Any as a pragmatic type hint

python
from typing import Any
def model(dbt: Any, session: Any):
...
Option 3: Inspect the actual class at runtime (for exploration)

python
def model(dbt, session):
print(type(dbt)) # reveals the generated class name
print(dir(dbt)) # shows available methods: ref, source, config, etc.
What methods dbt exposes:

dbt.ref("model_name") → returns a DataFrame/relation
dbt.source("source_name", "table_name") → same
dbt.config.get("param") → access model config vars
For Snowflake (Snowpark) specifically, you can type the session:

python
from snowflake.snowpark.session import Session
from typing import Any
def model(dbt: Any, session: Session) -> "Session.DataFrame":
dbt.config(materialized="table")
return session.table(dbt.ref("upstream_model"))
There's an open issue in dbt-core to expose proper stubs/type definitions — worth upvoting if you'd like to see official typing support!

**Comment by mateuscarestiato** [comment]
Hi @adrien-turiot-maxa! The dbt object in Python models is a dynamically generated class at runtime, which is why there's no obvious public type to import directly. Here are the practical solutions:

Option 1: Use TYPE_CHECKING (recommended)

python
from __future__ import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from dbt.adapters.protocol import MacroProtocol
def model(dbt: "MacroProtocol", session):
    df = dbt.ref("my_model")
    ...
This gives you IDE autocompletion without a runtime import error.

Option 2: Use Any as a pragmatic type hint

python
from typing import Any
def model(dbt: Any, session: Any):
    ...
Option 3: Inspect the actual class at runtime (for exploration)

python
def model(dbt, session):
    print(type(dbt))  # reveals the generated class name
    print(dir(dbt))   # shows available methods: ref, source, config, etc.
What methods dbt exposes:

dbt.ref("model_name") → returns a DataFrame/relation
dbt.source("source_name", "table_name") → same
dbt.config.get("param") → access model config vars
For Snowflake (Snowpark) specifically, you can type the session:

python
from snowflake.snowpark.session import Session
from typing import Any
def model(dbt: Any, session: Session) -> "Session.DataFrame":
    dbt.config(materialized="table")
    return session.table(dbt.ref("upstream_model"))
There's an open issue in dbt-core to expose proper stubs/type definitions — worth upvoting if you'd like to see official typing support!
